### ISSUE TITLE
Provide easily testable API for deprecating other APIs

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2919]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2919|
|Type|enhancement|
|Created|2007-11-20 22:38:23Z|
|Branch|https://github.com/twisted/twisted/tree/deprecation-api-2919|

The initial API should provide support for marking functions and methods as deprecated.

Users of the API shouldn't have to think too hard (or at all) about what the deprecation warning string is. 

Trial should grow something like `assertDeprecated` or `callDeprecated` that call deprecated code and checks that warnings are emitted. 

The API should probably interact with the release management version stuff.

Attachments:

* [2919.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/525/5253762a6393e8c01c11e27e1682756e65dd4436/500d0f4ff697c59be50d20e75881a1cbe1c6b7e4.diff) (2077 bytes) - added by therve on 2008-01-28 09:17:45Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2919 2919
type__enhancement enhancement
reporter__jml jml
priority__highest highest
milestone__ 
branch__branches_deprecation_api_2919 branches/deprecation-api-2919
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1195598303000000 1195598303000000
changetime__1201744275000000 1201744275000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

(In [22315]) Branching to 'deprecation-api-2919'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

I've done a prototype of an API for controlling which names a module or package exports.

It's here:  https://code.launchpad.net/~glyph/+junk/pyexport

Right now it doesn't deal with deprecation at all, but it does deal with "internal" names, which are similar.  (The big difference being that internal names just shouldn't be used at all, but in the case of deprecated names, you should be using something else instead.)

It also deals only with names that are at the top level of a module; however, controlling which methods _classes_ export is on my list of things to do, and similarly has some overlap with deprecation.

I wouldn't want to block this ticket on the export stuff being done or ready for inclusion in Twisted, but it would be good if this took it into account somehow.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @jml|
|-|-|

There are several failures with python2.3, but mainly this one I think:

```
[FAIL]: twisted.python.test.test_deprecate.TestDeprecationWarnings.test_deprecatedPreservesName

Traceback (most recent call last):
  File "/home/th/twisted_dev/twisted/trunk/twisted/python/test/test_deprecate.py", line 68, in test_deprecatedPreservesName
    self.assertEqual(self.dummyCallable.__name__, dummy.__name__)
twisted.trial.unittest.FailTest: not equal:
a = 'dummyCallable'
b = 'deprecatedFunction'
```

Pyflakes:
```
twisted/trial/unittest.py:21: 'getVersionString' imported but unused
```

In test_deprecate:
 * backslash must not be used to break import lines
 * the tests docstrings would benefit from a few `L{}`

In deprecate:
 * getWarningMethod and setWarningMethod are not tested directly. Overall, I don't think it's a great thing. Why don't you use warn_explicit?

One thing I wanted to point out is that we should keep the ```deprecated``` name for a more generic goal. We probably want to deprecate attributes and modules in the future, and even if we shouldn't do this here, I think deprecated should have another name (deprecatedMethod? deprecatedCallable? I'm not sure for now).

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml set owner to @therve|
|-|-|

I think that the getWarningMethod stuff is simple and clear and way better than monkey patching. If you want, you can paste a diff that uses warn_explicit and I can have a look.

I want to keep the name 'deprecated', future implementations can dispatch based on the type of the object.

I've done everything else.

I'll fix the 2.3 stuff a bit later.

Thanks for the review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

I have attached a simple diff removing getWarningMethod.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml commented|
|-|-|

That patch removes our ability to test the stacklevel of the warning.

In the past, we have mandated unit tests for the stacklevel of deprecation warnings. We need to test this in whatever shortcuts we provide.

**Comment by twisted-trac** [comment]
|<img alt="jml-insecure's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|jml-insecure set owner to @therve|
|-|-|

Fixed the 2.3 stuff. It turns out that `mergeFunctionMetadata` was horribly broken on Python 2.3. I've fixed this and made the tests a little clearer.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner|
|-|-|

Thanks for fixing mergeFunctionMetadata. It looks mostly good.

Regarding the test of the stacklevel, the problem with your code is that it forces the value to 2, whereas we can use some other levels in the code. I guess a third person will have to look at this.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv set owner to spiv<br>spiv set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|spiv set owner to @jml<br>spiv set status to `new`|
|-|-|

Why is getVersionString a function in the deprecations module, rather than a
method on Version?

There's multiple definitions of dummyCallable in test_deprecate.  Make a base
`TestCase` with it as a helper, or (probably better) just make it a
module-global function.

```
+    def test_deprecated_in_8_0(self):
+        """
+        The L{deprecated_in_8_0} decorator indicates that the callable was
+        deprecated in Twisted 8.0.
+        """
```

It really smells that deprecated_in_8_0 is defined in t.p.deprecate, even though
it's just for tests.

```
+class TestAppendToDocstring(TestCase):
```

This class ought to have a docstring like “Test _appendToDocstring function.”

```
+        def singleLineDocstring():
+            """This doesn't comply with standards, but is here for a test."""
+
```

You already defined this as a method on the `TestCase`.  Ditto multiLineDocstring.

There should probably be an `__all__` defined in twisted.python.deprecate.

```
+    def test_mergedFunctionBehavesLikeMergeTargetWithoutClosures(self):
+        # Now one without a closure
+        # XXX - I don't know what this test is for - jml, 2008-01-29.
```

I agree.  Delete it.

The twisted.python.deprecate module's docstring should include a brief guide
indicating how to use the contents of the module.

```
+class TestCallDeprecated(unittest.TestCase):
```

As usual, a simple docstring for the `TestCase` saying “```Tests for L{TestCase.callDeprecated} method.```” would be good.  (It's amazingly easy to
forget that new readers might not immediately know what sort of thing
“CallDeprecated” in the class name might be.)

```
+    def test_failsWithIncorrectDeprecation(self):
+        """
+        callDeprecated raises a test failure if the callable was deprecated
+        at a different version to the one expected.
+        """
+        differentVersion = Version('Foo', 1, 2, 3)
+        exception = self.assertRaises(
+            self.failureException,
+            self.callDeprecated,
+            differentVersion, self.oldMethod, 'foo')
+        self.assertIn(getVersionString(self.version), str(exception))
+        self.assertIn(getVersionString(differentVersion), str(exception))
```

I'm not certain this is a good idea.  I strongly suspect there will be tests
that may wish to use a deprecated function (or a function that calls a
deprecated function), but don't care about the details of the deprecation.

In fact, I'd expect tests that care about specific deprecations to be much rarer
than those that just wish to test functionality that is deprecated.  It seems
like having to suddenly get and pass the version in every test when a module is
deprecated will be a much larger burden than necessary.

To put it another way, I think callDeprecated taking and checking a version is a
YAGNI.

```
+    def test_nestedDeprecation(self):
+        """
+        L{callDeprecated} ignores all deprecations apart from the first.
+
+        We can expect the first to be the appropriate one, as deprecations are
+        expected to come first in a function implementation. Doing the
+        deprecation with a decorator is the same.
+        """
+
+        def nestedDeprecation(*args):
+            return self.oldMethod(*args)
+        nestedDeprecation = deprecated(self.version)(nestedDeprecation)
+
+        self.callDeprecated(self.version, nestedDeprecation, 24)
```

I don't understand this test.  Certainly I don't see how the nested deprecation
is any more or less ignored than the outer deprecation.  I'm not sure if the
docstring or test code or both is wrong.  Perhaps they are correct, but just
terribly unclear.

Finally, have you looked at `bzrlib.symbol_versioning`?  It seems a shame to
invent a similar-but-different API for the same thing.

**Comment by twisted-trac** [comment]
|<img alt="jml-insecure's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|jml-insecure set owner to spiv|
|-|-|

Replying to [spiv](#note_11):
> Why is getVersionString a function in the deprecations module, rather than a
> method on Version?
>

Moved.

> There's multiple definitions of dummyCallable in test_deprecate.  Make a base
> `TestCase` with it as a helper, or (probably better) just make it a
> module-global function.
>

Made a function.

> ```
> +    def test_deprecated_in_8_0(self):
> +        """
> +        The L{deprecated_in_8_0} decorator indicates that the callable was
> +        deprecated in Twisted 8.0.
> +        """
> ```
>
> It really smells that deprecated_in_8_0 is defined in t.p.deprecate, even though
> it's just for tests.
>

Removed.

> ```
> +class TestAppendToDocstring(TestCase):
> ```
>
> This class ought to have a docstring like “Test _appendToDocstring function.”
>

Added.

> ```
> +        def singleLineDocstring():
> +            """This doesn't comply with standards, but is here for a test."""
> +
> ```
>
> You already defined this as a method on the `TestCase`.  Ditto multiLineDocstring.
>

Removed.

> There should probably be an `__all__` defined in twisted.python.deprecate.
>

Yes.


> ```
> +    def test_mergedFunctionBehavesLikeMergeTargetWithoutClosures(self):
> +        # Now one without a closure
> +        # XXX - I don't know what this test is for - jml, 2008-01-29.
> ```
>
> I agree.  Delete it.
>

Deleted.

> The twisted.python.deprecate module's docstring should include a brief guide
> indicating how to use the contents of the module.
>

Added.

> ```
> +class TestCallDeprecated(unittest.TestCase):
> ```
>
> As usual, a simple docstring for the `TestCase` saying “```Tests for L{TestCase.callDeprecated} method.```” would be good.  (It's amazingly easy to
> forget that new readers might not immediately know what sort of thing
> “CallDeprecated” in the class name might be.)
>

Aded.

> ```
> +    def test_failsWithIncorrectDeprecation(self):
> +        """
> +        callDeprecated raises a test failure if the callable was deprecated
> +        at a different version to the one expected.
> +        """
> +        differentVersion = Version('Foo', 1, 2, 3)
> +        exception = self.assertRaises(
> +            self.failureException,
> +            self.callDeprecated,
> +            differentVersion, self.oldMethod, 'foo')
> +        self.assertIn(getVersionString(self.version), str(exception))
> +        self.assertIn(getVersionString(differentVersion), str(exception))
> ```
>
> I'm not certain this is a good idea.  I strongly suspect there will be tests
> that may wish to use a deprecated function (or a function that calls a
> deprecated function), but don't care about the details of the deprecation.
>
> In fact, I'd expect tests that care about specific deprecations to be much rarer
> than those that just wish to test functionality that is deprecated.  It seems
> like having to suddenly get and pass the version in every test when a module is
> deprecated will be a much larger burden than necessary.
>
> To put it another way, I think callDeprecated taking and checking a version is a
> YAGNI.
>

I based this on Twisted current practice and on Bazaar's framework. If people want to use deprecated functions without caring about the details, they can use the existing warning suppressors or add their own helper function to the Twisted base `TestCase`.

> ```
> +    def test_nestedDeprecation(self):
> +        """
> +        L{callDeprecated} ignores all deprecations apart from the first.
> +
> +        We can expect the first to be the appropriate one, as deprecations are
> +        expected to come first in a function implementation. Doing the
> +        deprecation with a decorator is the same.
> +        """
> +
> +        def nestedDeprecation(*args):
> +            return self.oldMethod(*args)
> +        nestedDeprecation = deprecated(self.version)(nestedDeprecation)
> +
> +        self.callDeprecated(self.version, nestedDeprecation, 24)
> ```
>
> I don't understand this test.  Certainly I don't see how the nested deprecation
> is any more or less ignored than the outer deprecation.  I'm not sure if the
> docstring or test code or both is wrong.  Perhaps they are correct, but just
> terribly unclear.
>

I hope that the changes I've made have made this clearer.

> Finally, have you looked at `bzrlib.symbol_versioning`?  It seems a shame to
> invent a similar-but-different API for the same thing.
>

I have. There are differences enough that a re-implementation is justified.
 * We have qual, and don't need to distinguish between functions and methods.
 * We are MIT, they are GPL
 * We have a version framework, they use strings.


Thanks for the review.

jml

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|spiv set owner to @jml|
|-|-|

Looks fine to me.  Merge it.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_14"></a>|@jml set status to `closed`|
|-|-|

(In [22428]) Deprecation API for functions and methods.

 * Author: jml
 * Reviewers: therve, spiv
 * Fixes [#2919](https://github.com/twisted/twisted/issues/2919)

Adds a deprecation decorator that can be used to easily and cleanly mark 
callables as deprecated.

This patch also fixes the `mergeFunctionMetadata` function to work in Python 
2.3.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
